### ISSUE TITLE
Remote stream not showing in RTCView

### ISSUE THREAD

**Issue by oktaysenkan** [react-native-webrtc/react-native-webrtc#774]
Hi, i need help for viewing remote stream. I can provide peer connection and i can see remote stream instance object on console but not showing RTCView. I using [react-native-kurento-utils](https://github.com/phoeshow/react-native-kurento-utils). I have custom turn server (coturn).

#### Emulator
![](https://i.imgur.com/gQ18Jz3.png)

#### onAddStream
![](https://i.imgur.com/6Emoj7r.png)

#### Remote stream object
![](https://i.imgur.com/ChZI2Nb.png)

#### React render
![](https://i.imgur.com/7oscLdt.png)

#### Kurento configs
![](https://i.imgur.com/gj3jhJ6.png)

**Comment by danjenkins** [comment]
The emulator is shite for video - use a real device and see if that fixes it.

**Comment by oktaysenkan** [comment]
> The emulator is shite for video - use a real device and see if that fixes it.

It's worked, thank you.

**Comment by max8392** [comment]
I have same problem, You have find resolve ?

**Comment by marwa-alnaqar121** [comment]
my video still in waiting don't connect could any one help me where's the problem
import React, {useState, useRef, useEffect, useCallback} from 'react';
import {useTheme, useNavigation} from '@react-navigation/native';
import {
  Image,
  StyleSheet,
  View,
  PanResponder,
  Animated,
  Dimensions,
  TouchableOpacity,
  I18nManager,
  ImageBackground,
} from 'react-native';
import io from 'socket.io-client';
import {
  mediaDevices,
  RTCIceCandidate,
  RTCPeerConnection,
  RTCSessionDescription,
  RTCView,
} from 'react-native-webrtc';
import Video from './Components/Video';
import InCallManager from 'react-native-incall-manager';
import RNVoipCall from 'react-native-voip-call';
import {v4 as uuidv4} from 'uuid';

/** components */
import {
  Text,
  Button,
  Container,
  SafeAreaView,
  RatingDialog,
} from '@components/common';
import I18n from '@i18n/I18n';
import RatingAlert from '@components/alert/RatingAlert';

/** redux */
import {useSelector, useDispatch} from 'react-redux';

import {Colors, Images, Metrics, Styles} from '@src/themes';
const AnimatedRTCView = Animated.createAnimatedComponent(RTCView);

/** data */
import {stums} from '@src/data/stums';

/** Config  stun */
const pc_config = {
  iceServers: [
    // {
    //   urls: 'stun:[STUN_IP]:[PORT]',
    //   'credentials': '[YOR CREDENTIALS]',
    //   'username': '[USERNAME]'
    // },
    {
      urls: ['turn:13.40.102.239:3478?transport=udp'],
      username: 'consultme',
      credential: 'consultme',
    },
    {
      urls: ['stun:stun.l.google.com:19302', 'stun:stun2.l.google.com:19302'],
    },
  ],
};

const SOCKET_SERVER_URL = 'sio.consultmekw.com';

/** Api */
import NotificationApi from '@api/NotificationApi';
import BookingApi from '@src/api/BookingApi';

const CallScreen = ({
  route: {
    params: {data, time = -1},
  },
}) => {
  const {
    colors: {primaryAll, white, TextBlack},
  } = useTheme();

  const navigation = useNavigation();

  const isLogin = useSelector(state => state.user.isLogin);
  const currentUser = useSelector(state => state.user.currentUser);
  const defaultLanguage = useSelector(state => state.lang.currentLang);
  const settingsContent = useSelector(state => state.settings.settingsContent);

  const socketRef = useRef();
  const pcsRef = useRef({});
  const localVideoRef = useRef(null);
  const localStreamRef = useRef();
  const [users, setUsers] = useState([]);
  const [stream, setStream] = useState(null);

  const [isFront, setIsFront] = useState(true);
  const [isMirror, setMirror] = useState(false);
  const [isMuted, setisMuted] = useState(false);
  const [idCall, setidCall] = useState('');
  const [isSpeaker, setisSpeaker] = useState(true);
  const [isVideoOff, setisVideoOff] = useState(false);
  const [globalTimeLeft, setGlobalTimeLeft] = useState(time);
  const [dropZoneValues, setDropZoneValues] = useState(null);
  const [UUID_CALL, setUUID_CALL] = useState(uuidv4);
  const [countdown, setCountDown] = useState('');

  /** alert */
  const [alertVisible, setAlertVisible] = React.useState(false);
  const [alertTitle, setAlertTitle] = React.useState('');
  const [alertMessage, setAlertMessage] = React.useState('');
  const [alertTextAction, setAlertTextAction] = React.useState('');
  const [alertTextCancel, setAlertTextCancel] = React.useState('');
  const [alertOnPress, setAlertOnPress] = React.useState(() => () => {});
  const [alertOnCancel, setAlertOnCancel] = React.useState(() => () => {});

  /** call */
  const [callClose, setCallClose] = React.useState(false);
  const [callStopScreen, setStopScreen] = React.useState(false);
  const [callStop, setCallStop] = React.useState(false);
  const [callStart, setCallStart] = React.useState(false);

  const countdownTime = useRef(0);

  useEffect(() => {
    InCallManager.start({media: 'video', auto: true});
    InCallManager.setKeepScreenOn(true);
    InCallManager.setForceSpeakerphoneOn(true);
    InCallManager.setSpeakerphoneOn(true);
    call();
    return () => {
      InCallManager.stop();
      addHistorical('1', 'stop');
    };
  }, []);

  const addHistorical = (connect, action) => {
    let send = true;
    if (action == 'stop' && callStop == false) {
      setStopScreen(true);
    } else if (action == 'stop call' && callStop == false) {
      setCallStop(true);
    } else if (action == 'start call' && callStart == false) {
      setCallStart(true);
    }

    if (send) {
      BookingApi.addHistorical(
        data?.id,
        data?.type_id?.id,
        currentUser?.id,
        countdownTime.current,
        connect,
        action,
      )
        .then(data => {
          if (action == 'stop') {
            setStopScreen(true);
          } else if (action == 'stop call') {
            setCallStop(true);
          } else if (action == 'start call') {
            setCallStart(true);
          }
        })
        .catch(error => {});
    }
  };

  const call = () => {
    //ADD HISTORICAL (START CALL)
    addHistorical('1', 'start call');
    if (data.action != 'Incoming call') {
      let token = data.type_id.device_token.device_id;
      let action = 'Incoming call';
      let title = 'Incoming call';
      let body = '';
      let id = data?.id;
      let date = data?.date;
      let time = data?.time;
      let duration = data?.duration;
      let memeber_id = currentUser?.id;
      let fname = currentUser?.fname;
      let lname = currentUser?.lname;
      let lawyer_id = data?.type_id?.id;
      let name = data?.type_id?.name;
      let name_ar = data?.type_id?.name_ar;
      let profession = data?.type_id?.profession;
      let profession_ar = data?.type_id?.profession_ar;
      if (data?.type_id?.device_token?.type === 'ios') {
        setTimeout(() => {
          NotificationApi.sendiOSCall(
            token,
            action,
            title,
            body,
            id,
            date,
            time,
            duration,
            memeber_id,
            fname,
            lname,
            lawyer_id,
            name,
            name_ar,
            profession,
            profession_ar,
            UUID_CALL,
          )
            .then(res => {
              setTimeout(() => {
                NotificationApi.sendNotifyiOS(data.type_id.id)
                  .then(res => {
                    console.log('success sendNotify');
                  })
                  .catch(error => {
                    console.log('error');
                  });
              }, 100);
            })
            .catch(error => {});
        }, 300);
      } else {
        setTimeout(() => {
          NotificationApi.sendNotifyAndroid(
            data?.type_id?.id,
            token,
            action,
            title,
            body,
            id,
            date,
            time,
            duration,
            memeber_id,
            fname,
            lname,
            lawyer_id,
            name,
            name_ar,
            profession,
            profession_ar,
            UUID_CALL,
          )
            .then(res => {
              console.log('success sendNotify');
            })
            .catch(error => {
              console.log('error');
            });
        }, 300);
      }
    } else {
      //ADD HISTORICAL (Join CALL)
      addHistorical('1', 'join call');
    }
  };

  const stop_call = () => {
    //ADD HISTORICAL (STOP CALL)
    addHistorical('0', 'stop call');
    if (data.action != 'Incoming call') {
      let token = data.type_id.device_token.device_id;
      let action = 'Missed call';
      let title = currentUser.lname + ' ' + currentUser.fname;
      let body = '';
      let id = data.id;
      let date = data.date;
      let time = data.time;
      let duration = data.duration;
      let memeber_id = '';
      let fname = '';
      let lname = '';
      let lawyer_id = data.type_id.id;
      let name = data.type_id.name;
      let name_ar = data.type_id.name_ar;
      let profession = data.type_id.profession;
      let profession_ar = data.type_id.profession_ar;

      if (data?.type_id?.device_token?.type === 'ios') {
        setTimeout(() => {
          NotificationApi.sendiOSCall(
            token,
            action,
            title,
            body,
            id,
            date,
            time,
            duration,
            memeber_id,
            fname,
            lname,
            lawyer_id,
            name,
            name_ar,
            profession,
            profession_ar,
            UUID_CALL,
          )
            .then(res => {})
            .catch(error => {});
        }, 300);
      } else {
        setTimeout(() => {
          NotificationApi.sendNotifyAndroid(
            data?.type_id?.id,
            token,
            action,
            title,
            body,
            id,
            date,
            time,
            duration,
            memeber_id,
            fname,
            lname,
            lawyer_id,
            name,
            name_ar,
            profession,
            profession_ar,
            UUID_CALL,
          )
            .then(res => {
              console.log('success sendNotify');
            })
            .catch(error => {
              console.log('error');
            });
        }, 300);
      }
    }
  };

  const getLocalStream = useCallback(async () => {
    // start every thing request permission to access to video and audio streams

    try {
      const localStream = await mediaDevices.getUserMedia({
        audio: true,
        video: {
          width: {min: 640, ideal: 1280, max: 1920},
          height: {min: 480, ideal: 720, max: 1080},
          minFrameRate: 30,
          facingMode: isFront ? 'user' : 'environment',
        },
      });
      //   getelementby id  and assign srcObject =localStream this will be enough to display my camera
      localStreamRef.current = localStream;
      //   set the strewm
      setStream(localStream);
      if (localVideoRef.current) {
        localVideoRef.current.setNativeProps({streamURL: localStream.toURL()});
      }
      if (!socketRef.current) return;
      const moonLanding = new Date();

      socketRef.current.emit('join_room', {
        room: data.id,
        email: moonLanding.getTime() + '@gmail.com',
      });
      console.log(`Join OK: `);
    } catch (e) {
      console.log(`getUserMedia error: ${e}`);
    }
  }, []);
  //  webrtc connection https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection
  const createPeerConnection = useCallback((socketID, email) => {
    try {
      // create connention and put inside it config of stun server "ice server"
      const pc = new RTCPeerConnection(pc_config);

      pc.onicecandidate = e => {
        if (!(socketRef.current && e.candidate)) return;
        console.log('onicecandidate', e.candidate);
        socketRef.current.emit('candidate', {
          candidate: e.candidate,
          candidateSendID: socketRef.current.id,
          candidateReceiveID: socketID,
        });
      };
      pc.oniceconnectionstatechange = e => {
        console.log(e);
      };
      pc.onaddstream = e => {
        setUsers(oldUsers =>
          oldUsers
            .filter(user => user.id !== socketID)
            .concat({
              id: socketID,
              email,
              stream: e.stream,
            }),
        );
      };

      if (localStreamRef.current) {
        console.log('localstream add');
        console.log(localStreamRef.current);
        console.log(localStreamRef.current._tracks);
        localStreamRef.current._tracks.forEach(track => {
          console.log('#####@track' + JSON.stringify(track));
          if (!localStreamRef.current) return;
          try {
            pc.addStream(localStreamRef.current);
          } catch (e) {
            console.log('addTrack' + e);
          }
        });
      } else {
        console.log('no local stream');
      }

      return pc;
    } catch (e) {
      console.error(e);
      return undefined;
    }
  }, []);

  useEffect(() => {
    socketRef.current = io.connect(settingsContent?.server);
    getLocalStream();
    socketRef.current.on('all_users', allUsers => {
      allUsers.forEach(async user => {
        if (!localStreamRef.current) return;
        const pc = createPeerConnection(user.id, user.email);
        if (!(pc && socketRef.current)) return;
        pcsRef.current = {...pcsRef.current, [user.id]: pc};
        // create offer " inner face store all info between us and remote peer and give pinch of methode " and
        try {
          const localSdp = await pc.createOffer({
            offerToReceiveAudio: true,
            offerToReceiveVideo: true,
          });
          //   finish create offer
          console.log('create offer success');
          //   trigger ice canidad to start pc.onicecandidate
          await pc.setLocalDescription(new RTCSessionDescription(localSdp));

          socketRef.current.emit('offer', {
            sdp: localSdp,
            offerSendID: socketRef.current.id,
            offerSendEmail: 'offerSendSample@sample.com',
            offerReceiveID: user.id,
          });
        } catch (e) {
          console.error(e);
        }
      });
    });

    socketRef.current.on('getOffer', async data => {
      const {sdp, offerSendID, offerSendEmail} = data;
      console.log('get offer');
      if (!localStreamRef.current) return;
      const pc = createPeerConnection(offerSendID, offerSendEmail);
      if (!(pc && socketRef.current)) return;
      pcsRef.current = {...pcsRef.current, [offerSendID]: pc};
      try {
        await pc.setRemoteDescription(new RTCSessionDescription(sdp));
        console.log('answer set remote description success');
        const localSdp = await pc.createAnswer({
          offerToReceiveVideo: true,
          offerToReceiveAudio: true,
        });
        await pc.setLocalDescription(new RTCSessionDescription(localSdp));
        socketRef.current.emit('answer', {
          sdp: localSdp,
          answerSendID: socketRef.current.id,
          answerReceiveID: offerSendID,
        });
      } catch (e) {
        console.error(e);
      }
    });

    socketRef.current.on(
      'getAnswer',
      (data: {sdp: RTCSessionDescription, answerSendID: string}) => {
        const {sdp, answerSendID} = data;
        console.log('get answer');
        const pc: RTCPeerConnection = pcsRef.current[answerSendID];
        if (!pc) return;
        pc.setRemoteDescription(new RTCSessionDescription(sdp));
      },
    );

    socketRef.current.on('getCandidate', async data => {
      console.log('get candidate');
      const pc = pcsRef.current[data.candidateSendID];
      if (!pc) return;
      await pc.addIceCandidate(new RTCIceCandidate(data.candidate));
      console.log('candidate add success');
    });

    socketRef.current.on('user_exit', data => {
      if (!pcsRef.current[data.id]) return;
      pcsRef.current[data.id].close();
      delete pcsRef.current[data.id];
      setUsers(oldUsers => oldUsers.filter(user => user.id !== data.id));
      RNVoipCall.endAllCalls();
      stop_call();
      openPopup(
        I18n.t('review.Rate lawyer'),
        '',
        I18n.t('review.reviews'),
        I18n.t('review.Not now'),
        () => {
          closePopup();
        },
        () => {
          closePopup();
        },
      );
    });

    return () => {
      if (socketRef.current) {
        socketRef.current.disconnect();
      }
      users.forEach(user => {
        if (!pcsRef.current[user.id]) return;
        pcsRef.current[user.id].close();
        delete pcsRef.current[user.id];
      });
    };
  }, [createPeerConnection, getLocalStream]);

  const _renderVideo = (obj, index, nbr) => {
    return (
      <View
        key={index}
        style={
          nbr == 1
            ? {
                width: Metrics.deviceWidth,
                height: Metrics.deviceHeight,
              }
            : {
                width: Metrics.deviceWidth / 2,
                height:
                  nbr === 2 || nbr === 1
                    ? Metrics.deviceHeight / 2
                    : Metrics.deviceHeight / Math.round(nbr / 2) - 40,
              }
        }>
        {index === 0 && (
          <View
            style={{
              position: 'absolute',
              zIndex: 1,
              top: 0,
              left: 0,
              width: Metrics.deviceWidth / 2,
              height: Metrics.deviceHeight / 2 - 40,
              backgroundColor:
                isVideo === false ? Colors.black : Colors.transparent,
              //justifyContent: 'flex-end',
              //alignItems: 'flex-end',
              //flexDirection: 'row',
              //...Styles.center,
              borderWidth: isVideo === false ? 1 : 0,
              borderColor:
                isVideo === false ? Colors.white50 : Colors.transparent,
            }}>
            <View
              style={{
                flex: 1,
                ...Styles.center,
              }}>
              {isVideo === false && (
                <View style={styles.containerIconBig}>
                  <Image
                    resizeMode={'contain'}
                    style={styles.iconBig}
                    source={Images.camera.disable_camera}
                  />
                </View>
              )}
            </View>
            <View
              style={{
                ...Styles.rowAlign,
              }}>
              {isMuted && (
                <View style={styles.containerIcon}>
                  <Image
                    resizeMode={'contain'}
                    style={styles.icon}
                    source={Images.camera.disable_speaker}
                  />
                </View>
              )}
              {isSpeaker === false && (
                <View style={styles.containerIcon}>
                  <Image
                    resizeMode={'contain'}
                    style={styles.icon}
                    source={Images.camera.disable_speaker}
                  />
                </View>
              )}
            </View>
          </View>
        )}
        <RTCView
          muted={true}
          streamURL={obj}
          style={{
            flex: 1,
          }}
          mirror={true}
          objectFit={'cover'}
        />
      </View>
    );
  };

  let listUser = users.map(user => user.stream.toURL());

  const pan = useRef(
    new Animated.ValueXY({
      x: Metrics.deviceWidth - 25 - 120,
      y: Metrics.deviceHeight - 180 - 25,
    }),
  ).current;
  //to start to drag only after a 10% drag threshold
  const DRAG_THRESHOLD = Dimensions.get('screen').height * 0.1;

  //to limit the drag on 80% of the screen height
  const DRAG_LIMIT = Dimensions.get('screen').height * 0.8;
  const panResponder = useRef(
    PanResponder.create({
      onMoveShouldSetPanResponder: () => true,
      onPanResponderGrant: () => {
        pan.setOffset({
          x: pan.x._value,
          y: pan.y._value,
        });
      },
      onPanResponderMove: Animated.event([null, {dx: pan.x, dy: pan.y}]),
      /* onPanResponderMove: (e, gesture) => {
              console.log(gesture);
      
              //for THRESHOLDs
              if ((Math.abs(gesture.dy) > DRAG_THRESHOLD) &&
                (Math.abs(gesture.dy) < DRAG_LIMIT)) {
                return Animated.event([
                  null, { dx: pan.x, dy: pan.y }
                ])(e, gesture)
              }
            }, */

      onPanResponderRelease: () => {
        pan.flattenOffset();
      },
    }),
  ).current;

  const muteMicrophone = () => {
    localStreamRef?.current?.getAudioTracks()?.forEach(track => {
      track.enabled = !track.enabled;
      setisMuted(!track.enabled);
    });
  };

  const stopCamera = () => {
    localStreamRef?.current?.getVideoTracks().forEach(track => {
      track.enabled = !track.enabled;
      setisVideoOff(!track.enabled);
    });
  };

  const switchSpeaker = () => {
    InCallManager.setForceSpeakerphoneOn(!isSpeaker);
    InCallManager.setSpeakerphoneOn(!isSpeaker);
    InCallManager.setKeepScreenOn(!isSpeaker);
    setisSpeaker(!isSpeaker);
  };

  const mirrorCamera = () => {
    setMirror(!isMirror);
  };

  const switchCamera = () => {
    localStreamRef?.current?.getVideoTracks().forEach(track => {
      track._switchCamera();
      setIsFront(last => !last);
    });
  };

  const Countdown = seconds => {
    const [timeLeft, setTimeLeft] = useState(seconds);
    const intervalRef = useRef(); // Add a ref to store the interval id

    useEffect(() => {
      intervalRef.current = setInterval(() => {
        setTimeLeft(t => t - 1);
      }, 1000);
      return () => clearInterval(intervalRef.current);
    }, []);

    // Add a listener to `timeLeft`
    useEffect(() => {
      if (timeLeft <= 0) {
        clearInterval(intervalRef.current);
        if (timeLeft <= 0) {
          RNVoipCall.endAllCalls();
          stop_call();
          openPopup(
            I18n.t('review.Rate lawyer'),
            '',
            I18n.t('review.reviews'),
            I18n.t('review.Not now'),
            () => {
              closePopup();
            },
            () => {
              closePopup();
            },
          );
        }
      }
      countdownTime.current = (time - timeLeft) / 60;
    }, [timeLeft]);
    return (
      <Text
        color={timeLeft < 60 ? Colors.red.C0 : Colors.white}
        fontStyle={[20]}>
        {callClose ? '' : new Date(timeLeft * 1000).toISOString().substr(11, 8)}
      </Text>
    );
  };

  const openPopup = (
    title,
    message,
    textAction,
    textCancel,
    onPress,
    onCancel,
  ) => {
    setAlertVisible(true);
    setAlertTitle(title);
    setAlertMessage(message);
    setAlertTextAction(textAction);
    setAlertTextCancel(textCancel);
    setAlertOnPress(() => onPress);
    setAlertOnCancel(() => onCancel);
  };

  const closePopup = () => {
    setAlertVisible(false);
    setTimeout(() => {
      navigation.goBack();
    }, 800);
  };

  return (
    <SafeAreaView style={{flex: 1, backgroundColor: Colors.primary}}>
      <View style={{flex: 1}}>
        {
          <View
            style={[
              {
                zIndex: 4,
              },
              users.length == 0 && {
                flex: 1,
                //height: 0
              },
              users.length == 1 && {
                position: 'absolute',
                right: 25,
                top: 50,
                backgroundColor: 'black',
              },
              users.length == 1 && styles.box,
            ]}>
            <View
              style={{
                overflow: 'hidden',
                backgroundColor: Colors.black.color,
                position: 'relative',
                zIndex: 1,
                flex: 1,
              }}>
              <RTCView
                style={[
                  users.length == 0 && {
                    flex: 0.95,
                  },
                  users.length == 1 && styles.box,
                  isVideoOff && {
                    flex: 0,
                    height: 0.1,
                  },
                ]}
                zOrder={1}
                muted={false}
                mirror={isFront}
                ref={localVideoRef}
                autoPlay
                objectFit={'cover'}></RTCView>
              {isVideoOff && (
                <View
                  style={{
                    flex: 1,
                    ...Styles.center,
                    backgroundColor: Colors.black.color,
                  }}>
                  <Image
                    resizeMode={'contain'}
                    style={{width: 30, height: 30, tintColor: Colors.white}}
                    source={Images.camera.disable_camera}
                  />
                </View>
              )}
            </View>
          </View>
        }
        {users.map((user, index) => (
          <Video
            key={index}
            email={user.email}
            stream={user.stream}
            style={{flex: 1}}
          />
        ))}
      </View>
      <View
        style={{
          width: Metrics.deviceWidth,
          height: 150,
          position: 'absolute',
          zIndex: 10,
          bottom: (Metrics.deviceWidth * 210) / 600 + 20,
          paddingHorizontal: 30,
        }}>
        <Container isFlex backgroundColor={Colors.transparent}>
          <Text color={Colors.white} fontStyle={[20, 'bold']}>
            {I18nManager.isRTL ? data.type_id.name_ar : data.type_id.name}
          </Text>
          {users.length === 0 && (
            <Text color={Colors.white} fontStyle={[14, 'bold']}>
              {I18n.t('call.Waiting') + ' ...'}
            </Text>
          )}
          <Text
            numberOfLines={2}
            color={Colors.white}
            fontStyle={[14]}
            style={{width: 200}}>
            {I18nManager.isRTL
              ? data.type_id.profession_ar
              : data.type_id.profession}
          </Text>
        </Container>
        <View
          style={{
            backgroundColor: Colors.white20,
            borderRadius: 30 / 2,
            width: 120,
            height: 30,
            ...Styles.rowAlign,
            ...Styles.center,
          }}>
          <View
            style={{
              width: 12,
              height: 12,
              borderRadius: 6,
              backgroundColor: 'red',
              marginRight: 5,
            }}
          />
          {Countdown(globalTimeLeft)}
        </View>
      </View>
      <ImageBackground
        style={{
          width: Metrics.deviceWidth,
          height: (Metrics.deviceWidth * 210) / 600,
          tintColor: Colors.white,
          position: 'absolute',
          zIndex: 10,
          bottom: 0,
        }}
        source={Images.images.bottom}>
        <View
          style={{
            flex: 1,
            paddingTop: 10,
          }}>
          <Container isCenter isFlex backgroundColor={Colors.transparent}>
            <TouchableOpacity
              onPress={() => {
                RNVoipCall.endAllCalls();
                stop_call();
                setCallClose(true);
                if (socketRef.current) {
                  socketRef.current.disconnect();
                }
                users.forEach(user => {
                  if (!pcsRef.current[user.id]) return;
                  pcsRef.current[user.id].close();
                  delete pcsRef.current[user.id];
                });
                openPopup(
                  I18n.t('review.Rate lawyer'),
                  '',
                  I18n.t('review.reviews'),
                  I18n.t('review.Not now'),
                  () => {
                    closePopup();
                  },
                  () => {
                    closePopup();
                  },
                );
              }}
              style={{
                borderRadius: 62 / 2,
                width: 62,
                height: 62,
                ...Styles.center,
                backgroundColor: 'red',
                marginTop: 10,
              }}>
              <Image
                source={Images.call.phone_}
                style={{width: 32, height: 32, tintColor: Colors.white}}
              />
            </TouchableOpacity>
          </Container>
          <Container
            isFlex
            backgroundColor={Colors.transparent}
            isRowCenter
            isSpaceBetween
            hasPaddingHorizontal
            style={{marginBottom: 15}}>
            <Container isRowCenter backgroundColor={Colors.transparent}>
              <TouchableOpacity
                onPress={stopCamera}
                style={{
                  width: 50,
                  height: 50,
                  ...Styles.center,
                  borderRadius: 50 / 2,
                  backgroundColor: Colors.secondary + (isVideoOff ? '8F' : ''),
                  marginHorizontal: Metrics.baseMargin,
                }}>
                <Image
                  source={
                    isVideoOff ? Images.call.no_camera : Images.call.camera
                  }
                  style={{width: 20, height: 20, tintColor: Colors.white}}
                />
              </TouchableOpacity>
              <TouchableOpacity
                onPress={muteMicrophone}
                style={{
                  width: 50,
                  height: 50,
                  ...Styles.center,
                  borderRadius: 50 / 2,
                  backgroundColor: Colors.secondary + (isMuted ? '8F' : ''),
                }}>
                <Image
                  source={
                    isMuted ? Images.call.no_microphone : Images.call.microphone
                  }
                  style={{width: 20, height: 20, tintColor: Colors.white}}
                />
              </TouchableOpacity>
            </Container>
            <Container isRowCenter backgroundColor={Colors.transparent}>
              <TouchableOpacity
                onPress={switchSpeaker}
                style={{
                  width: 50,
                  height: 50,
                  ...Styles.center,
                  borderRadius: 50 / 2,
                  backgroundColor: Colors.secondary + (isSpeaker ? '' : '8F'),
                  marginHorizontal: Metrics.baseMargin,
                }}>
                <Image
                  source={
                    isSpeaker ? Images.call.speaker : Images.call.no_speaker
                  }
                  style={{width: 20, height: 20, tintColor: Colors.white}}
                />
              </TouchableOpacity>
              <TouchableOpacity
                onPress={switchCamera}
                style={{
                  width: 50,
                  height: 50,
                  ...Styles.center,
                  borderRadius: 50 / 2,
                  backgroundColor: Colors.secondary,
                }}>
                <Image
                  source={
                    isSpeaker
                      ? Images.call.camera_rotation
                      : Images.call.camera_rotation
                  }
                  style={{width: 22, height: 22, tintColor: Colors.white}}
                />
              </TouchableOpacity>
            </Container>
          </Container>
        </View>
      </ImageBackground>
      <RatingAlert
        user_id={currentUser.id}
        lawyer_id={data.type_id.id}
        booking_id={data.id}
        lang={defaultLanguage}
        show={alertVisible}
        showProgress={false}
        title={alertTitle}
        message={alertMessage}
        closeOnTouchOutside={true}
        closeOnHardwareBackPress={false}
        showCancelButton={true}
        showConfirmButton={true}
        cancelText={alertTextCancel}
        confirmText={alertTextAction}
        confirmButtonColor={Colors.primary}
        cancelButtonColor={Colors.grey.C9}
        onConfirmPressed={alertOnPress}
        onCancelPressed={alertOnCancel}
        onClosePressed={() => {
          closePopup();
        }}
        contentContainerStyle={{backgroundColor: white}}
        titleStyle={{color: TextBlack}}
        messageStyle={{color: TextBlack}}
      />
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  sectionContainer: {
    marginTop: 32,
    paddingHorizontal: 24,
  },
  sectionTitle: {
    fontSize: 24,
    fontWeight: '600',
  },
  sectionDescription: {
    marginTop: 8,
    fontSize: 18,
    fontWeight: '400',
  },
  highlight: {
    fontWeight: '700',
  },

  content: {
    flex: 1,
    backgroundColor: Colors.white,
    marginHorizontal: Metrics.doubleBaseMargin,
  },
  image: {
    width: Metrics.deviceWidth / 2,
    height: (Metrics.deviceHeight - 45) / 3,
  },
  footer: {
    position: 'absolute',
    zIndex: 1,
    bottom: 10,
  },
  containerIcon: {
    ...Styles.center,
    margin: Metrics.baseMargin,
    backgroundColor: Colors.white50,
    width: 30,
    height: 30,
    borderRadius: 15,
  },
  icon: {
    tintColor: Colors.white,
    width: 16,
    height: 16,
  },
  containerIconBig: {
    ...Styles.center,
    margin: Metrics.baseMargin,
    backgroundColor: Colors.white50,
    width: 45,
    height: 45,
    borderRadius: 45 / 2,
  },
  iconBig: {
    tintColor: Colors.white,
    width: 20,
    height: 20,
  },
  box: {
    height: 180,
    width: 120,
    overflow: 'hidden',
  },
});

export default CallScreen;
